src/nomad_simulations/properties/band_gap.py

```diff
@@ -0,0 +1,147 @@
+        """,
+    )
+
+    momentum_transfer = Quantity(
+        type=np.float64,
+        shape=[2, 3],
+        description="""
+        If the `ElectronicBandGap` is `'indirect'`, the reciprocal momentum transfer for which the band gap is defined
+        in units of the `reciprocal_lattice_vectors`. The initial and final momentum 3D vectors are given in the first
+        and second element. Example, the momentum transfer in bulk Si2 happens between the Γ and the (approximately)
+        X points in the Brillouin zone; thus:
+            `momentum_transfer = [[0, 0, 0], [0.5, 0.5, 0]]`.
+
+        Note: this quantity only refers to scalar `value`, not to arrays of `value`.
+        """,
+    )
+
+    spin_channel = Quantity(
+        type=np.int32,
+        description="""
+        Spin channel of the corresponding `ElectronicBandGap`. It can take values of 0 or 1.
+        """,
+    )
+
+    value = Quantity(
+        type=np.float64,
+        unit='joule',
+        description="""
+        The value of the `ElectronicBandGap`.
+        """,
+    )
+
+    def __init__(self, m_def: Section = None, m_context: Context = None, **kwargs):
+        super().__init__(m_def, m_context, **kwargs)
+        self.name = self.m_def.name
+        self.rank = []  # ? Is this here or in the attrs instantiation better?
+
+    def check_negative_values(self, logger: BoundLogger) -> pint.Quantity:
+        """
+        Checks if the electronic band gaps is negative and sets them to 0 if they are.
+
+        Args:
+            logger (BoundLogger): The logger to log messages.
+        """
+        value = self.value.magnitude
+        if not isinstance(self.value.magnitude, np.ndarray):  # for scalars
+            value = np.array([value])
+
+        # Set the value to 0 when it is negative
+        if (value < 0).any():
+            logger.warning(
+                'The electronic band gap cannot be defined negative. We set them up to 0.'
+            )
+        value[value < 0] = 0
+
+        if not isinstance(self.value.magnitude, np.ndarray):  # for scalars
+            value = value[0]
+        return value * self.value.u
+
+    def resolve_type(self, logger: BoundLogger) -> Optional[str]:
+        """
+        Resolves the `type` of the electronic band gap based on the stored `momentum_transfer` values.
+
+        Args:
+            logger (BoundLogger): The logger to log messages.
+
+        Returns:
+            (Optional[str]): The resolved `type` of the electronic band gap.
+        """
+        if (
+            self.momentum_transfer is None or len(self.momentum_transfer) < 2
+        ) and self.type == 'indirect':
+            logger.warning(
+                "The `momentum_transfer` is not properly defined for an `type='indirect'` electronic band gap."
+            )
+            return None
+        if self.momentum_transfer is not None and len(self.momentum_transfer) > 0:
```

Comment:
No, it is taken care of having a defined length for later operations

Reply:
You mean for `np.isclose(momentum_difference, np.zeros(3))`? Still, the logic is kind of clunkily formulated.
You can just roll up both `if`-statements into one

```python
if self.momentum_transfer is not None or len(self.momentum_transfer) < 2:
    logger.warning(
                    'The `momentum_transfer` should have at least two elements so that the difference can be calculated and the type of electronic band gap can be resolved.'
                )
                return None
```

Then the rest of the code no longer has to live under the 1st `if`-statement.
Moreover, I think you can further tighten the 2nd conditional to `len(self.momentum_transfer) != 2`.

Reply:
Ok, I changed a bit the logic to make it more understandable.